Bluetooth: GATT: Make bt_gatt_unsubscribe work while disconnected

Vudentz opened the pull request:
This allows application to call bt_gatt_unsubscribe while disconnected
as otherwise the parameters cannot be cleared/resused until a device is
finally connected again.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4234,6 +4230,11 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
 	params->value = 0x0000;
 
+	/* Don't attempt to write if not connected */
+	if (conn->state != BT_CONN_CONNECTED) {
```
I'm wondering if the unsubscribe should be queued, in the same way as a service changed would be. 
Unsubscribing this way now would mean that the peer would never be informed, and could just keep sending notifications that would be silently dropped.

Vudentz replied:
We would need to add memory storage for unsubscribing it so the argument would be then if it is worth the extra memory, also note that we do have a similar problem with BT_GATT_SUBSCRIBE_FLAG_VOLATILE as it does unsubscribe automatically as well, since this is on the client side I don't think it breaks anything but perhaps you want it to be stricly used with unpair?

carlescufi replied:
What's the point of unsubscribing without letting the peer know? The app could simply drop the notifications instead once it reconnects?

Vudentz committed:
Bluetooth: ISO: Fix revert direction when setting SDU

s_sdu refers to rx/input and m_sdu refers to tx/output.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Fix not calling channel disconnected

When the channel has just been bound but is not connected yet there is
no need to send any command over the air but the disconnected callback
shall still be called in order to notify the channel owner that it has
reached disconnected state.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Fix cleanup connection

This fixes bt_iso_cleanup when there are still channels bound to the
ACL connection.

On top of it introduce bt_iso_chan_unbind which can be used to unbind
channels and thus release the reference to the ACL connection if that
has not been disconnected in which case the channels are unbind
automatically.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Use CHECKIF instead of __ASSERT_NO_MSG

CHECKIF is preferred as it handles full runtime error handling in
addition to just asserts.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: conn: Fix unbalance reference when an ISO has been connected

This fixes a regression introduced by 4350021f09c which prevents
references to be properly release when there is an ISO connection.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: conn: Fix invalid state transition

When initiating a disconnection conn_disconnect would be called to send
HCI_Disconnect but the controller may be quick enough to send Disconnect
Complete event before setting BT_CONN_DISCONNECT which will then cause
the invalid transition from BT_CONN_DISCONNECTED to BT_CONN_DISCONNECT
and the connection won't be freed properly.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Fix not unbind if ACL has been disconnected

If ACL has been disconnected any ISO handle shall be considered invalid
as well.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
